jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java

```diff
@@ -0,0 +1,343 @@
+                return false;
+            }
+
+            return true;
+        }
+
+        /**
+         * Returns the start vertex of the path
+         *
+         * @return the start vertex of the path
+         */
+        public V getHead() {
+            return vertices.get(0);
+        }
+
+        /**
+         * Returns the end vertex of the path
+         *
+         * @return the end vertex of the path
+         */
+        public V getTail() {
+            return vertices.get(vertices.size() - 1);
+        }
+
+        /**
+         * Returns whether the path is empty, i.e. has no vertices
+         *
+         * @return whether the path is empty
+         */
+        public boolean isEmpty() {
+            return vertices.isEmpty();
+        }
+
+        /**
+         * Returns an ordered list of the vertices of the path
+         *
+         * @return an ordered list of the vertices of the path
+         */
+        public List<V> getVertices() {
+            return vertices;
+        }
+
+        /**
+         * Returns the labels of all vertices of the path
+         *
+         * @return the labels of all vertices of the path
+         */
+        public Set<Integer> getLabels() {
+            return labels;
+        }
+
+        /**
+         * Returns the cost of the path
+         *
+         * @return the cost of the path
+         */
+        public double getCost() {
+            return cost;
+        }
+
+        /**
+         * Returns a shallow copy of this labeled path instance. Vertices are not cloned.
+         *
+         * @return a shallow copy of this path.
+         *
+         * @throws RuntimeException in case the clone is not supported
+         *
+         * @see java.lang.Object#clone()
+         */
+        public LabeledPath<V> clone() {
+            try {
+                LabeledPath<V> newLabeledPath = TypeUtil.uncheckedCast(super.clone());
+                newLabeledPath.vertices = (ArrayList<V>) this.vertices.clone();
+                newLabeledPath.labels = new HashSet<>(this.labels);
+                newLabeledPath.cost = this.cost;
+
+                return newLabeledPath;
+            } catch (CloneNotSupportedException e) {
+                e.printStackTrace();
+                throw new RuntimeException();
+            }
+        }
+    }
+
+    private Graph<V, E> graph;
+    private Map<V, Integer> labels;
+    private int lengthBound;
+
+    /**
+     * Constructs an algorithm with given inputs
+     *
+     * @param graph the (improvement) graph on which to calculate the local augmentation
+     * @param lengthBound the upper bound for the length of cycles to detect
+     * @param labels the labels of the vertices encoding the subsets of vertices
+     */
+    public AhujaOrlinSharmaCyclicExchangeLocalAugmentation(Graph<V, E> graph, int lengthBound, Map<V, Integer> labels) {
+        this.graph = Objects.requireNonNull(graph, "Graph cannot be null");
```

Comment:
Are there any other restrictions on the input graph, e.g. directed vs undirected?  Are multiple edges between vertex pairs allowed?  Doesn't the graph need to be weighted?